egs/wsj/s5/steps/nnet3/components.py

```diff
@@ -83,20 +83,51 @@ def AddOutputNode(config_lines, input, label_delay=None):
 def AddFinalLayer(config_lines, input, output_dim, ng_affine_options = "", label_delay=None, include_softmax = "true"):
     prev_layer_output = AddAffineLayer(config_lines, "Final", input, output_dim, ng_affine_options)
     if include_softmax == "true":
-      prev_layer_output = AddSoftmaxLayer(config_lines, "Final", prev_layer_output)
+        prev_layer_output = AddSoftmaxLayer(config_lines, "Final", prev_layer_output)
     AddOutputNode(config_lines, prev_layer_output, label_delay)
 
+# the following two functions are necessary for state preserving training. Users need to add additional output-nodes 
+# in order to output states of the recurrent nodes that would be used as inputs to the next minibatch.
+# We use the string "_STATE_PREVIOUS_MINIBATCH" as a reserved suffix for node names: If there is an output-node named "xxx"
+# other than "output", there should be an input-node named "xxx_STATE_PREVIOUS_MINIBATCH". On the other hand, if there is
```

Comment:
Yes thanks for pointing it out.  I should have had a more general assumption that the network could have multiple outputs with costs. Will also modify Nnet::Check() accordingly to allow it (at Ln 688 in nnet-nnet.cc).